### PR TITLE
Save startup timeline

### PR DESCRIPTION
**Pull request by liyuqian**
This would help us investigate issues like
https://github.com/flutter/flutter/issues/64781

### CODE REVIEW

**Comment by liyuqian** [comment]
@jonahwilliams : any suggestion on how to unit test this? `packages/flutter_tools/test` doesn't seem to have anything related with `--trace-startup`.

**Comment by zanderso** [comment]
The `Tracing` class looks like it needs some love before it can be tested properly. In particular, the instance method `downloadStartupTrace()` creates a new instance?

https://github.com/flutter/flutter/blob/cb819504f3388383ae9a2deeccb5c2d6f0dfd25f/packages/flutter_tools/lib/src/tracing.dart#L97

@jonahwilliams thoughts?

**Comment by jonahwilliams** [comment]
There is a FakeVmServiceHost which can be configured for a set of request/responses, there should be a number of tests that use that pattern. You should be able to provide that fake vm service in place of the real one to get a unit test going.

**Comment by jonahwilliams** [comment]
For example: https://github.com/flutter/flutter/blob/9b7b9d795edfdd12c6ad8dd6495bed789c1bac05/packages/flutter_tools/test/general.shard/coverage_collector_test.dart#L12

**Review by darielopper — APPROVED** [review]

**Comment by jonahwilliams** [comment]
@dramosv90 I believe we asked you several months ago to stop approving PRs that you have not really reviewed, or are not involved in the design decision for.

**Comment by liyuqian** [comment]
@jonahwilliams : test added :)

File: packages/flutter_tools/test/general.shard/tracing_test.dart [hunk 1]
```diff
@@ -234,4 +234,85 @@ void main() {
...
+    // Validate that old tracing data is deleted.
+    final File timelineFile = fileSystem.currentDirectory.childFile('start_up_timeline.json')
+      ..writeAsStringSync('stale');
+
+    await downloadStartupTrace(fakeVmServiceHost.vmService,
+      output: fileSystem.currentDirectory,
+      logger: logger,
+    );
+
+    final dynamic kExpectedTimeline = json.decode('''
```
Comment (jonahwilliams) [inline packages/flutter_tools/test/general.shard/tracing_test.dart:285]: if you parse the result and compare a dart map, the test will be resilient to changes in formatting.

**Review by jonahwilliams — APPROVED** [review]
LGTM with nit

File: packages/flutter_tools/test/general.shard/tracing_test.dart (hunk 1 above)
Comment (liyuqian) [reply packages/flutter_tools/test/general.shard/tracing_test.dart]: Done.

**Comment by zanderso** [comment]
What's the status of this PR? Do the presubmit checks need another kick?

**Comment by liyuqian** [comment]
"Google tests" seems to be stuck. Merge manually now.

**Pull request merged**